src/Microsoft.AspNet.Mvc.Abstractions/ModelBinding/ValueProviderResult.cs

```diff
@@ -142,9 +142,17 @@ private object ConvertSimpleType(CultureInfo culture, object value, Type destina
             if (!(canConvertFrom || converter.CanConvertTo(destinationType)))
             {
                 // EnumConverter cannot convert integer, so we verify manually
-                if (destinationType.GetTypeInfo().IsEnum && (value is int))
+                if (destinationType.GetTypeInfo().IsEnum &&
+                    (value is int ||
+                    value is uint ||
+                    value is long ||
+                    value is ulong ||
+                    value is short ||
+                    value is ushort ||
+                    value is byte ||
+                    value is sbyte))
                 {
-                    return Enum.ToObject(destinationType, (int)value);
+                    return Enum.ToObject(destinationType, Convert.ToInt32(value));
```

Comment:
The loss of precision here concerns me. Can we add a unit test for really, really, *really* big long values? I.e. `long.MaxValue`?

The check for `IsDefined` on the other hand does not concern me. A validator should do that work, not type conversion.